Issue#190 Change query parameters mapping in presentation layer

tex1988 opened the pull request:
dev
## ZenHub

* [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009/issues/ita-social-projects/home/190)


## Code reviewers

- [ ] @kh0ma, @VadymSokorenko, @denis-litvinov, @DzigMS

### Second Level Review

- [ ] @github_username

## Summary of issue

Optimization of query parameters mapping in home-application module

## Summary of change

Added specialized QueryApiService for mapping query parameters with using jax-rs UriInfo interface

## Testing approach

Automated

## CHECK LIST
- [x]  СI passed
- [x]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

tex1988 committed:
Issue#190: changed manual filterMap formation with UriInfo

tex1988 committed:
Issue#190: Added common generalized interface QueryableService for entity services
Added QueryApiService interface and implementation to hide query logic from presentation layer
Changed Api implementations to use QueryApiService

tex1988 committed:
Issue#190: Changed EXCLUDED_PARAMS_SET name to EXCLUDED_PARAMETERS in QueryApiService
Changed EXCLUDED_PARAMETERS initialization from manual adding parameters with a stream in QueryApiService
Some code formatting in QueryApiService
Added javadoc in QueryApiService, QueryableService and QueryApiServiceImpl

tex1988 committed:
Issue#190: Fixed import order according to project style in QueryApiService
Fixed javadoc for getFilterMap method in QueryApiService

tex1988 committed:
Issue#190: Replaced getPageNumber, getPageSize, getFilter, getSort with unified static method getParameterValue in QueryApiService
Changed getFilterMap from default to static in QueryApiService
Added default implementation of getPageFromQuery method in QueryApiService
Added getSpecification method in QueryApiService

tex1988 committed:
Issue#190: EntitySpecificationService with all query configs moved to home-application module

tex1988 committed:
Issue#190: EntitySpecificationService class renamed to EntitySpecificationServiceImpl
Added EntitySpecificationService interface
Changed filter argument type from Map to MultivaluedMap in getSpecification method of EntitySpecificationServiceImpl class
Changed toRSQLString method on EntitySpecificationServiceImpl to use MultivaluedMap instead of Map
Changed return type of getFilterMap method from Map to MultivaluedMap in QueryApiService
Changed EntitySpecificationService dependency from implementation to interface in CommonApi

tex1988 committed:
Issue#190: Changed imports order according to project style

tex1988 committed:
Issue#190: Added javadoc for getSpecification method in EntitySpecificationService

tex1988 committed:
Issue#190: Improved description in javadoc for getSpecification method in EntitySpecificationService

tex1988 committed:
Issue#190: Created query package in config package of home-application module
Moved rsql package into query package
Moved query config classes into query package
Moved HomeUserDetailsService from config package into service package

tex1988 committed:
Issue#190: Added protected modifier for buildQueryResponse method in CommonApi
Synced with changes in Issue#196 Fix contact type query

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/query/rsql/UserRSQLEndpointConfig.java:
```diff
@@ -21,8 +21,8 @@
-        map.put(UserQueryParamEnum.LAST_NAME, "lastName");
+        map.put(UserQueryParamEnum.FIRST_NAME, "first_name");
+        map.put(UserQueryParamEnum.LAST_NAME, "last_name");
         map.put(UserQueryParamEnum.CONTACT_PHONE, "contacts.phone");
```
Sorry, it isn't connected to your changes/story but the value of this put is wrong It must be the name of a query parameter from spec. The same is true for the line below. 

Could you please change it to the proper value:
```java
 map.put(UserQueryParamEnum.CONTACT_PHONE, "contact_phone")
...
```
```

tex1988 replied:
Proposed changes have been made

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/exception/mapper/ExceptionMapperUtils.java:
```diff
@@ -15,7 +15,7 @@ private ExceptionMapperUtils() {
     public static String getBadRequestExceptionMessageParser(RSQLParserException exception) {
         String message = null;
-        if (!exception.getMessage().contains("<EOF>")) {
+        if (!exception.getMessage().contains("<EOF>") && !exception.getMessage().contains("\" <OR> \", \"\"")) {
```
What is the purpose of this addition? How is it connected to your story?

tex1988 replied:
Changes rewerted

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/QueryApiService.java:
```diff
@@ -0,0 +1,111 @@
+public interface QueryApiService<T extends BaseEntity, D extends BaseDto> {
+    Set<String> EXCLUDED_PARAMETERS = new HashSet<>(Arrays.stream(DefaultQueryParams.values())
+            .map(DefaultQueryParams::getParameter)
+            .collect(Collectors.toList()));
```
Can we use `Collectors.toSet()` instead of wrapping the list into Set?

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/QueryApiService.java:
```diff
@@ -0,0 +1,111 @@
+     * @return Map of query and path parameters required by EntitySpecificationService
+     * to build Filter specification
+     */
+    static MultivaluedMap<String, String> getFilterMap(UriInfo uriInfo) {
```
We also have to consider the situation when a query param is present in request and has the same name as a path param to avoid the following scenario:
`GET /users/52/contacts?user_id=51`
In this case, we'll get wrong unexpected results. I suggest throwing client error referring to wrong query param set.

Using more streams we can do it in this way =)

```java
    static MultivaluedMap<String, String> getFilterMap(UriInfo uriInfo) {
        return Stream.of(uriInfo.getPathParameters(), uriInfo.getQueryParameters())
                .flatMap(map -> map.entrySet().stream())
                .filter(not(entry -> EXCLUDED_PARAMETERS.contains(entry.getKey())))
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        Map.Entry::getValue,
                        (v1, v2) -> {
                            throw new IllegalArgumentException(); //TODO Client error
                        },
                        MultivaluedHashMap::new
                ));
    }
```

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/QueryApiService.java:
```diff
@@ -0,0 +1,111 @@
+     *                  that provides access to application and request URI information
+     * @return String value of specified query parameter
+     */
+    static String getParameterValue(String param, UriInfo uriInfo) {
```
We also have to consider the situation when more than 1 special query param present in the request and validate it as a client error as cannot randomly take the parameter.

`/users?pageNum=2&pageNum=3&pageNum=4` - it should be validated complaining that only one `pageNum` can be present in the request.

Consider the following:

```java
    static String getParameterValue(String param, UriInfo uriInfo) {
        String paramValue = null;
        MultivaluedMap<String, String> queryParameters = uriInfo.getQueryParameters();
        if (queryParameters.containsKey(param)) {
            List<String> paramValues = queryParameters.get(param);
            switch (paramValues.size()) {
                case 0: // Is it possible for this Map?
                    paramValue = null;
                    break;
                case 1:
                    paramValue = paramValues.get(0);
                    break;
                default:
                    throw new IllegalArgumentException(); //TODO client error
            }
        }
        return paramValue;
    }
```

tex1988 replied:
Proposed changes have been made but in another way

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/impl/EntitySpecificationServiceImpl.java:
```diff
@@ -0,0 +1,46 @@
+
+    public static final String RSQL_EQUAL = "==";
+
+    public static final String RSQL_EQUAL_OR = "=in=";
```
Why do we need it? Seems like the change from not approved PR. Please avoid it. You should deliver only your changes as part of your PR.

tex1988 replied:
Changes rewerted

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/impl/EntitySpecificationServiceImpl.java:
```diff
@@ -0,0 +1,46 @@
+                .filter(entry -> entry.getValue() != null)
+                .map(entry -> entry.getValue().stream()
+                        .map(valueElement -> entry.getKey() + RSQL_EQUAL_OR
+                                + "(" + valueElement + "," + valueElement.toUpperCase() + ")")
```
The same

tex1988 replied:
Changes rewerted

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/impl/EntitySpecificationServiceImpl.java:
```diff
@@ -0,0 +1,46 @@
+@Service
+public class EntitySpecificationServiceImpl<T> implements EntitySpecificationService<T> {
+
+    public static final String DEFAULT_SORT = "id,desc";
```
Isn't default come from the controller? Do we still need it here?

tex1988 replied:
Proposed changes have been made

tex1988 committed:
Revert "Issue#190: Added protected modifier for buildQueryResponse method in CommonApi Synced with changes in Issue#196 Fix contact type query"

This reverts commit 13e85bef

tex1988 committed:
Revert "Issue#190: Changed CONTACT_PHONE and CONTACT_EMAIL values according to API spec in UserRSQLEndpointConfig

tex1988 committed:
Revert "Issue#190: Removed DEFAULT_SORT constant and logic expression in sortSpecification initialization of EntitySpecificationServiceImpl, as a default value comes from the controller

tex1988 committed:
Revert "Issue#190: Changed Collectors.toList() to Collectors.toSet() in EXCLUDED_PARAMETERS initialization of QueryApiService
Added validation for checking that query params has only one value in QueryApiService
Added BadRequestHomeException and BadRequestHomeExceptionMapper for proper information of API user about wrong request query

tex1988 committed:
Issue#190: Fixed import order in BadRequestHomeExceptionMapper according to project style rules

tex1988 committed:
Issue#190: Added missing @Provider annotation

tex1988 committed:
Issue#190: Changes in QueryApiService and QueryApiServiceImpl

Changed EXCLUDED_PARAMETERS from set to map to be able to provide default values of default parameters in QueryApiService
Added validation for same path and query parameters in QueryApiService
Added default values of parameters in DefaultQueryParams in QueryApiService
Changed return parameter type to Optional of getParameterValue method in QueryApiService
Changed getSpecification implementation in EntitySpecificationServiceImpl according to getParameterValue changes in QueryApiService

tex1988 committed:
Issue#190: Fixed javadoc in QueryApiService

tex1988 committed:
Merge branch 'dev' into Issue#190-Change-query-parameters-mapping-in-presentation-layer

tex1988 committed:
Issue#190: Changed assigning null value to filter variable in QueryApiServiceImpl

tex1988 committed:
Issue#190: Changed filterMap validation to queryParams validation in getFilterMap method to avoid validation skipping of default query parameters in QueryApiService

Abhai2016 commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/impl/QueryApiServiceImpl.java:
```diff
@@ -0,0 +1,32 @@
+ */
+@Service
+@RequiredArgsConstructor
+public class QueryApiServiceImpl<T extends BaseEntity, D extends BaseDto> implements QueryApiService<T, D> {
```
Not sure it's worth it to add services and apis for query just to get rid of filling map with query parameters because it just adds more code and makes it more complex which doesn't match with pattern KISS

kh0ma replied:
How many times did you make mistake filling this map? I found at least two places. The same can be with any dev. In case when we can avoid any manual routine operation we should automate it =)

kh0ma commented on home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/QueryApiService.java:
```diff
@@ -0,0 +1,128 @@
+     * @return Optional of String value of specified query parameter
+     */
+    static Optional<String> getParameterValue(String param, UriInfo uriInfo) {
+        if (uriInfo.getQueryParameters().get(param) != null) {
```
It could be applicable here but in general, try to keep your method to have a single entry and a single exit meaning only one return statement.

Together with Optional in your case could be achieved as the following:
```
return Optional.ofNullable(uriInfo.getQueryParameters().get(param))
    .map(<map to Iterable>)
    .map(<get next as first>);
```
